### PR TITLE
Prometheus: Kickstart your query, formerly query patterns

### PR DESCRIPTION
**Pull request by bohandley**
### Kickstart Your Query
fixes https://github.com/grafana/observability-metrics-squad/issues/49

#### What is this? 
Query patterns select is being replaced by the kickstart your query modal. This gives users a better understanding of what is happening when they chose a query pattern. Users can choose a query pattern, apply a query pattern to a written query, they can replace a query pattern and can add a new query pattern as an additional query.

Update Dec 28th: adding new query pattern type, binary query.

This is being changed
Screen Shot 2022-12-22 at 10 34 32 PM
to this, where a user clicks "Kick start your query"
Screen Shot 2022-12-22 at 10 35 01 PM
and this modal opens with query patterns.
Screen Shot 2022-12-22 at 10 35 16 PM

#### Why is this still a draft? 
Docs need to be edited/updated and there is a possibility of adding new query patterns as described [here](https://raintank-corp.slack.com/archives/C03045EE1EC/p1671003274693539). Query patterns need better descriptions as well.

Full proposal: https://docs.google.com/document/d/1PHEz6G7TkP0D4lF9c42bgFT9ukhXhmiN4-klUWOPN0o/edit?usp=sharing

### CODE REVIEW

**Review by itsmylife — COMMENTED** [review]
Could you please handle a11y requirements? Like adding some aria tags to make things more accessible

**Comment by itsmylife** [comment]
Also so far looks good. But I have one question.
In the `code` mode when I pick a kickstart query it gives me the barebones. In the builder mode that is enough because you **see** what's missing. But in the `code` mode you don't see them. Which is understandable. If somebody is using `code` mode he might have some idea about metrics and labels. But still, can we at least add template of the query in the modal we show the queries?

Like instead of showing this:
```
avg(sum by() (rate([$__rate_interval])))
```
we can show this:
```
avg(sum by(<metric_label>) (rate(<metric_name>{<metric_label>="metric_label_value"}[$__rate_interval])))
```
WDYT?

**Comment by bohandley** [comment]
> Could you please handle a11y requirements? Like adding some aria tags to make things more accessible

Can do!

**Comment by bohandley** [comment]
> Also so far looks good. But I have one question. In the `code` mode when I pick a kickstart query it gives me the barebones. In the builder mode that is enough because you **see** what's missing. But in the `code` mode you don't see them. Which is understandable. If somebody is using `code` mode he might have some idea about metrics and labels. But still, **can we at least add template of the query in the modal we show the queries**?
> 
> Like instead of showing this:
> 
> ```
> avg(sum by() (rate([$__rate_interval])))
> ```
> 
> we can show this:
> 
> ```
> avg(sum by(<metric_label>) (rate(<metric_name>{<metric_label>="metric_label_value"}[$__rate_interval])))
> ```
> 
> WDYT?

I think we can show it in the modal! I'll get working on that.

**Comment by itsmylife** [comment]
> I think it is helpful, but I also think this does not look that great in the code. WDYT? It might look better if the templates for metric and labels was italicized but I don't think that is possible. Or maybe this does look ok?

This should be enough. When somebody clicks the "Use this Query" button we will show the barebones as we have it right now.

**Comment by gtk-grafana** [comment]
First off, what a cool feature! I've got a couple of notes/questions.

image

In the kickstart your query I'm not seeing the current metric name getting pulled into the kickstart your query modal, since the figma designs are for Loki I can't exactly tell if this is desired functionality or not. I would assume that we would pull in the metric name that is in the query if it's available?

image
But then if we make that change we'll need to adjust the styles since some metric names might be longer then the available space.

Does this functionality compete with the query hints that show up when you select a counter or other metric type that triggers this UI?
image
Do we want to keep the existing functionality, or change that link to pull up the new modal? Or is this functionality supposed to work alongside that feature?

File: docs/sources/datasources/prometheus/query-editor/index.md [hunk 1]
```diff
@@ -203,13 +202,6 @@ To add the operations to your query, click the hint.
 
 Explain mode helps you understand a query by displaying a step-by-step explanation of all query components and operations.
 
-### Raw query
-
-{{< figure src="/static/img/docs/prometheus/raw-query-8-5.png" max-width="500px" class="docs-image--no-shadow" caption="Raw query" >}}
```
Comment (gtk-grafana) [inline docs/sources/datasources/prometheus/query-editor/index.md:208]: good catch

File: public/app/plugins/datasource/prometheus/querybuilder/QueryPattern.tsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  const lang = { grammar: promqlGrammar, name: 'promql' };
+
+  return (
+    <Card className={styles.card}>
+      <Card.Heading>{pattern.name}</Card.Heading>
+      <div className={styles.rawQueryContainer}>
+        <RawQuery
+          aria-label={`${pattern.name} raw query`}
+          query={promQueryModeller.renderQuery({
+            labels: [],
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/querybuilder/QueryPattern.tsx:36]: labels is supposed to be empty?

File: public/app/plugins/datasource/prometheus/querybuilder/QueryPattern.tsx [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+          className={styles.rawQuery}
+        />
+      </div>
+      <Card.Actions>
+        {selectedPatternName !== pattern.name ? (
+          <Button
+            size="sm"
+            aria-label="use this query button"
+            onClick={() => {
+              if (hasPreviousQuery) {
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/querybuilder/QueryPattern.tsx:50]: This hasPreviousQuery only seems to work when they've accepted a previous query pattern, would we want to show this whenever there's any text in the query editor (basically always)?

File: public/app/plugins/datasource/prometheus/querybuilder/QueryPatternsModal.test.tsx [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+import { QueryPatternsModal } from './QueryPatternsModal';
+import { PromQueryPatternType } from './types';
+
+// don't care about interaction tracking in our unit tests
+jest.mock('@grafana/runtime', () => ({
+  ...jest.requireActual('@grafana/runtime'),
+  reportInteraction: jest.fn(),
+}));
+
+const defaultProps = {
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/querybuilder/QueryPatternsModal.test.tsx:15]: great job adding tests!

**Review by gtk-grafana — APPROVED** [review]
This looks great, I've got a couple of questions on desired functionality, but everything is working for me locally and the code looks good.
LGTM

**Comment by bohandley** [comment]
> First off, what a cool feature! I've got a couple of notes/questions.
> 
> image
> 
> In the kickstart your query I'm not seeing the current metric name getting pulled into the kickstart your query modal, since the figma designs are for Loki I can't exactly tell if this is desired functionality or not. I would assume that we would pull in the metric name that is in the query if it's available?
> 
> image
> 
> But then if we make that change we'll need to adjust the styles since some metric names might be longer then the available space.
> Does this functionality compete with the query hints that show up when you select a counter or other metric type that triggers this UI? image Do we want to keep the existing functionality, or change that link to pull up the new modal? Or is this functionality supposed to work alongside that feature?

The Kick start functionality lives outside the actual query and is a hard coded config object that is applied to the query, so any previously chosen metrics wouldn't effect what is displayed in the modal. I think we might be able to populate the modal with previous query metrics/labels/values but you're right, the styling would need to be considered.

Competing with query hints, good eye! It kind of does!!! Query hints will add a specific function based on the metric(possibly for beginning Prom users), but the Kickstart query patterns are generic/general patterns regardless of metric (or the Prom user is an expert and knows they want to apply them quickly). So yeah! This needs more research! They're coming at the problem of making suggestions from different angles and so I think they are competing a little. I'll make a card to explore this more, maybe add some metrics/event tracking for this. Thanks Galen!

File: docs/sources/datasources/prometheus/query-editor/index.md (hunk 1 above)
Comment (bohandley) [reply docs/sources/datasources/prometheus/query-editor/index.md:208]: 🕺🏻 Thank you!

File: public/app/plugins/datasource/prometheus/querybuilder/QueryPattern.tsx (hunk 2 above)
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/querybuilder/QueryPattern.tsx:36]: Yep! Ismail suggested we fill them with templates in a comment above, and we can, but currently it looks a little busy. We can add labels here, but they would be just placeholders and not applied to the query.

File: public/app/plugins/datasource/prometheus/querybuilder/QueryPattern.tsx (hunk 3 above)
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/querybuilder/QueryPattern.tsx:50]: Good point! It does only work when they've accepted a previous query pattern and thinking about it, it doesn't actually replace the query, it only applies the chosen query pattern to what is already there! Like, it you choose sum then rate, it does that, then choose histogram to "replace" the query, you get a histogram of the sum and rate! You're right, if there is a query already entered, we should use this functionality, whether it be a metric or a previous query pattern. Also, I think we should change the wording to from "replace this query" to "apply this pattern."

File: public/app/plugins/datasource/prometheus/querybuilder/QueryPatternsModal.test.tsx (hunk 4 above)
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/querybuilder/QueryPatternsModal.test.tsx:15]: Thank you! 🥳

**Comment by bohandley** [comment]
> This looks great, I've got a couple of questions on desired functionality, but everything is working for me locally and the code looks good. LGTM

I think those were great comments and I will be making a change regarding the functionality around "hasPreviousQuery." Thank you @gtk-grafana!

**Comment by itsmylife** [comment]
This looks great and working well. It is already LGTM but I just want to see if you are going to add more here. So I am holding my approval for now :)

File: public/app/plugins/datasource/prometheus/querybuilder/QueryPattern.tsx (hunk 3 above)
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/querybuilder/QueryPattern.tsx:50]: I made the change, this now asks a user to apply the query pattern if there is a previous metric, label, operation or binary query in the editor whereas previously it only checked operations.

**Comment by bohandley** [comment]
> This looks great and working well. It is already LGTM but I just want to see if you are going to add more here. So I am holding my approval for now :)

@itsmylife I made some small updates based on @gtk-grafana 's suggestions. I think this is good to go!

**Review by itsmylife — APPROVED** [review]
🚀 🚀

**Comment by LifHkr** [comment]
Are there any plans to make this editable? We are using a very limited Prometheus data source that does not support rate and many other types of queries. Showing these query examples will further confuse the users if they are not valid. If not, would there be a way to disable this feature from appearing?

**Comment by gtk-grafana** [comment]
@LifHkr Currently we have no plan to make this editable, I think we'd need to understand more about this. If you create a [feature request](https://github.com/grafana/grafana/discussions/new?category=data-sources) and provide some details (data source type and version), we'll be more easily to understand your use-case.

**Pull request merged**